src/controllers/krate/versions.rs

```diff
@@ -249,131 +247,93 @@ async fn list_by_semver(
-            .load_stream::<(i32, String, bool)>(conn)
-            .await?
-            .try_for_each(|(id, num, yanked)| {
-                let semver = semver::Version::parse(&num).ok();
-                sorted_versions.insert(id, (semver, yanked, None));
-                future::ready(Ok(()))
-            })
-            .await?;
+    };
 
-        sorted_versions
-            .sort_unstable_by(|_, (semver_a, _, _), _, (semver_b, _, _)| semver_b.cmp(semver_a));
+    let mut query = make_base_query();
 
+    if let Some(options) = options {
         assert!(
             !matches!(&options.page, Page::Numeric(_)),
             "?page= is not supported"
         );
-
-        let release_tracks = include.release_tracks.then(|| {
-            ReleaseTracks::from_sorted_semver_iter(
-                sorted_versions
-                    .values()
-                    .filter(|(_, yanked, _)| !yanked)
-                    .filter_map(|(semver, _, _)| semver.as_ref()),
+        if let Some(SeekPayload::Semver(Semver { num, id })) = Seek::Semver.after(&options.page)? {
+            query = query.filter(
+                versions::semver_ord
+                    .eq(semver_ord(num.clone()))
+                    .and(versions::id.lt(id))
+                    .or(versions::semver_ord.lt(semver_ord(num))),
             )
-        });
-
-        let mut idx = Some(0);
-        if let Some(SeekPayload::Semver(Semver { id })) = Seek::Semver.after(&options.page)? {
-            idx = sorted_versions
-                .get_index_of(&id)
-                .filter(|i| i + 1 < sorted_versions.len())
-                .map(|i| i + 1);
         }
-        if let Some(start) = idx {
-            let end = (start + options.per_page as usize).min(sorted_versions.len());
-            let ids = sorted_versions[start..end]
-                .keys()
-                .cloned()
-                .collect::<Vec<_>>();
+        query = query.limit(options.per_page);
+    }
+
+    query = query.order((versions::semver_ord.desc(), versions::id.desc()));
+
+    let data: Vec<(Version, Option<User>)> = query.load(conn).await?;
+    let mut next_page = None;
+    if let Some(options) = options {
+        next_page = next_seek_params(&data, options, |last| Seek::Semver.to_payload(last))?
+            .map(|p| req.query_with_params(p));
+    };
+
+    let release_tracks = if params.include()?.release_tracks {
+        let mut sorted_versions = IndexSet::new();
+        if options.is_some() {
             versions::table
                 .filter(versions::crate_id.eq(crate_id))
-                .left_outer_join(users::table)
-                .select(<(Version, Option<User>)>::as_select())
-                .filter(versions::id.eq_any(ids))
-                .load_stream::<(Version, Option<User>)>(conn)
+                .filter(not(versions::yanked))
+                .select(versions::num)
+                .load_stream::<String>(conn)
```

Comment:
Is a `.order(versions::semver_ord.desc())` call missing here, or am I missing something?